blog/2024-03-15-reflex-architecture.md

```diff
@@ -0,0 +1,166 @@
+
+This also let's our users bring their own components if we don't have a component they need.
+
+### Backend (State)
+
+In Reflex only the frontend compiles to Javascript, while all the state and logic stays in Python and is run on the server.
+
+```python
+import reflex as rx
+
+class GithubState(rx.State):
+    username: str = "reflex-dev"
+
+    def set_username(self, username: str):
+        self.username = username
+```
+
+Normally when writing web apps, you have to write a lot of boilerplate code to connect the frontend and backend. With Reflex, you don't have to worry about that. We handle the communication between the frontend and backend for you.
+
+We follow a reactive approach to state. All the variables that change over time are defined in a `State` class. When a variable changes, the frontend is automatically updated to reflect the new state.
+
+
+## Event Processing
+
+Now we get into the interesting part - how we handle events and state updates. Let's walk through it with our Github profile image example.
+
+### Event Trigger
+
+Events are initiated when the user interacts with the UI. In our example, when the user types in the input field and then clicks away, the `on_blur` event is triggered.
+
+### Event Queue
+
+On the frontend, we maintain an event queue of all pending events. When an event is triggered, it is added to the queue. We have a `processing` flag to make sure only one event is processed at a time. This ensures that the state is always consistent and there aren't any race conditions with two event handlers modifying the state at the same time.
+
+Once the event is ready to be processed, it is sent to the backend through a websocket connection.
+
+### State Manager
+
+The event contains information including the token of the client, the event handler to run, and the arguments to pass to the event handler. The backend processes the event and updates the state.
+
+The first step is to get the user's state. This is where the state manager comes in. It is essentially a mapping from a user's token to their state. When the backend receives an event, it looks up the user's state and runs the event handler on that state.
+
+By default the state manager is just an in-memory dictionary, but it can be extended to use a database or cache. In production we use Redis to store the state.
+
+### Event Handling
+
+Once we have the user's state, the next step is to run the event handler.
+
+```python
+def set_username(self, username: str):
+    self.username = username
+```
+
+In our example, the `set_username` event handler is run on the user's state. This updates the user's state with the new username.
+
+### State Updates
+
+After the event handler runs, it updates the state's vars. Whenever a state var updates it, Reflex marks it as "dirty". When the event handler is done processing, we find all the dirty vars and send them to the frontend to update the UI. This allows our apps to scale as the state grows.
```

Comment:
first mention of `state vars` and no mention of what they are or how they fit into the rest of the story, they should be defined earlier in the text

Reply:
This line: `Whenever a state var updates it, Reflex marks it as "dirty". ` should change to `Whenever a state var is updated, Reflex marks it as "dirty". `

Reply:
added a note above defining vars